ni_measurementlink_service/_internal/parameter/_message.py

```diff
@@ -0,0 +1,94 @@
+import struct
+
+from google.protobuf.internal import encoder, wire_format
+
+
+def _inner_message_encoder(field_index):
+    """Mimics google.protobuf.internal.MessageEncoder.
+
+    See EncodeField:
+    https://github.com/protocolbuffers/protobuf/blob/0b817d46d4ca1977d3dccf2442aeee3c9e98e3a1/python/google/protobuf/internal/encoder.py#L765C15-L765C15
+    """
+    tag = encoder.TagBytes(field_index, wire_format.WIRETYPE_LENGTH_DELIMITED)
+    encode_varint = _varint_encoder()
+
+    def _encode_message(write, value, deterministic):
+        write(tag)
+        bytes = value.SerializeToString()
+        encode_varint(write, len(bytes), deterministic)
+        write(bytes)
+
+    return _encode_message
+
+
+def _varint_encoder():
+    """Return an encoder for a basic varint value (does not include tag).
+
+    From google.protobuf.internal.encoder.py _VarintEncoder
+    """
+    local_int2byte = struct.Struct(">B").pack
+
+    def encode_varint(write, value, unused_deterministic=None):
+        bits = value & 0x7F
+        value >>= 7
+        while value:
+            write(local_int2byte(0x80 | bits))
+            bits = value & 0x7F
+            value >>= 7
+        return write(local_int2byte(bits))
+
+    return encode_varint
+
+
+def _inner_message_decoder(field_index, is_repeated, is_packed, key, new_default):
+    """Based on google.protobuf.internal.MessageDecoder.
+
+    See DecodeField
+    """
+
+    def _convert_to_byte_string(memview):
+        byte_str = memview.tobytes()
+        return byte_str
+
+    def _decode_message(buffer, pos, end, message, field_dict):
+        decode_varint = _varint_decoder(mask=(1 << 64) - 1, result_type=int)
+        value = field_dict.get(key)
+        if value is None:
+            value = field_dict.setdefault(key, new_default(message))
+        # Read length.
+        (size, pos) = decode_varint(buffer, pos)
+        new_pos = pos + size
```

Comment:
You removed two error checks:
- Bounds check
- Did parsing consume the entire buffer subset?

Reply:
Added back in the checks. ParseFromString is not like _InternalSerialize and seems to return the size of bytes it parsed. I can't find where ParseFromString is in the code to confirm this.